core/BTB.hpp

```diff
@@ -0,0 +1,277 @@
+    private:
+        uint32_t index_mask_;
+        uint32_t index_shift_;
+    };
+
+    class BTB : public sparta::Unit
+    {
+    public:
+        static constexpr const char * name = "btb";
+
+        // Parameters
+        //! \brief Parameters for BTB
+        class BTBParameterSet : public sparta::ParameterSet
+        {
+        public:
+            BTBParameterSet(sparta::TreeNode* n) :
+                sparta::ParameterSet(n)
+            {}
+
+            PARAMETER(uint64_t, num_of_entries, 4096, "BTB # of entries (power of 2)")
+            PARAMETER(uint32_t, block_size, 64, "BTB Search stride in bytes (power of 2)")
+            PARAMETER(uint32_t, associativity, 8, "BTB associativity (power of 2)")
+        };
+
+
+        BTB(sparta::TreeNode * node, const BTBParameterSet *p) : sparta::Unit(node)
+        {
+            cache_.reset(new sparta::cache::Cache(p->num_of_entries,
+                                                  1,
+                                                  p->block_size,
+                                                  BTBEntry(),
+                                                  sparta::cache::TreePLRUReplacement(p->associativity),
+                                                  false));
+            cache_->setAddrDecoder(new BTBAddrDecoder(p->num_of_entries, p->block_size, p->associativity));
+        }
+
+        /**
+         *\return whether an addr is in the cache
+            */
+        bool isHit(uint64_t addr) const
+        {
+            const BTBEntry *line = cache_->peekItem(addr);
+            return ( line != nullptr );
+        }
+
+        // Get a line for replacement
+        BTBEntry &getLineForReplacementWithInvalidCheck(uint64_t addr)
+        {
+            return cache_->getCacheSet(addr).getItemForReplacementWithInvalidCheck();
+        }
+
+        /**
+         *\return Pointer to line with addr.  nullptr is returned if not fould
+        */
+        BTBEntry * getLine(uint64_t addr)
+        {
+            return cache_->getItem(addr);
+        }
+
+        /**
+         *\return Pointer to line with addr.  nullptr is returned if not fould
+        */
+        const BTBEntry * peekLine(uint64_t addr) const
+        {
+            return cache_->peekItem(addr);
+        }
+
+        void touchLRU(const BTBEntry &line)
+        {
+            auto *rep = cache_->getCacheSetAtIndex( line.getSetIndex() ).getReplacementIF();
+            rep->touchLRU( line.getWay() );
+        }
+
+        void touchMRU(const BTBEntry &line)
+        {
+            auto *rep = cache_->getCacheSetAtIndex( line.getSetIndex() ).getReplacementIF();
+            rep->touchMRU( line.getWay() );
+        }
+
+        // Allocate 'line' as having the new 'addr'
+        // 'line' doesn't know anything about NT
+        void allocateWithMRUUpdate(BTBEntry &line,
+                                    uint64_t   addr)
+        {
+            line.reset( addr );
+            touchMRU( line );
+        }
```

Comment:
Why does this take a `line` and then reset it?  How about it creates one and returns it?